azure-iot-device/azure/iot/device/common/pipeline/pipeline_thread.py

```diff
@@ -0,0 +1,199 @@
+    name.
+    """
+    executor = getattr(_get_named_executor, thread_name, None)
+    if not executor:
+        logger.info("Creating {} executor".format(thread_name))
+        executor = ThreadPoolExecutor(max_workers=1)
+        setattr(_get_named_executor, thread_name, ThreadPoolExecutor(max_workers=1))
+    return executor
+
+
+def _get_thread_local_storage():
+    """
+    Get an object which contains "thread local storage" using the threading.local()
+    function.  If no such object exists, one is created.  The object returned from
+    `threading.local()` appears to be a global object, but has a different value for
+    each individual thread.  We use this so we can check which thread we're running
+    inside at runtime.
+    """
+    if not getattr(_get_thread_local_storage, "local", None):
+        logger.info("Creating thread local storage")
+        _get_thread_local_storage.local = threading.local()
+    return _get_thread_local_storage.local
+
+
+def _invoke_on_executor_thread(thread_name, block=True, _func=None):
+    """
+    Decorator which runs the wrapped function on a given thread.  If block==False,
+    the call returns immediately without waiting for the decorated function to complete.
+    If block==True, the call waits for the decorated function to complete before returning.
+    """
+
+    def decorator(func):
+        function_name = getattr(func, "__name__", str(func))
+
+        def wrapper(*args, **kwargs):
+            if not getattr(_get_thread_local_storage(), "in_{}_thread".format(thread_name), False):
+                logger.info("Starting {} in {} thread".format(function_name, thread_name))
+
+                def thread_proc():
+                    setattr(_get_thread_local_storage(), "in_{}_thread".format(thread_name), True)
+                    return func(*args, **kwargs)
+
+                # TODO: add a timeout here and throw exception on failure
+                future = _get_named_executor(thread_name).submit(thread_proc)
+                if block:
+                    return future.result()
+                else:
+                    return future
+            else:
+                logger.debug("Already in {} thread for {}".format(thread_name, function_name))
+                return func(*args, **kwargs)
+
+        # Silly hack:  On 2.7, we can't use @functools.wraps on callables don't have a __name__ attribute
+        # attribute(like MagicMock object), so we only do it when we have a name.  functools.update_wrapper
+        # below is the same as using the @functools.wraps(func) decorator on the wrapper function above.
+        if getattr(func, "__name__", None):
+            return functools.update_wrapper(wrapped=func, wrapper=wrapper)
+        else:
+            return wrapper
+
+    if _func is None:
+        return decorator
+    else:
+        return decorator(_func)
+
+
+def invoke_on_pipeline_thread(_func=None):
```

Comment:
>_func=None [](start = 30, length = 10)

is the function truly optional ?

Reply:
Yes, if you're using this as a decorator, it gets called with _func = None.  If you use it to wrap a function (like foo = invoke_on_pipeline_thread(bar).(()), then _func = bar

---
In reply to: [301759558](https://github.com/Azure/azure-iot-sdk-python-preview/pull/133#discussion_r301759558) [](ancestors = 301759558)

Reply:
I thought decorators always took the function they are decorating as the first argument? Am I wrong about this?

---
In reply to: [302245692](https://github.com/Azure/azure-iot-sdk-python-preview/pull/133#discussion_r302245692) [](ancestors = 302245692,301759558)

Reply:
I was wrong about this.  I was (incorrectly) copying a pattern that I didn't need here.  
(The pattern that I didn't need is here: [https://realpython.com/primer-on-python-decorators/#both-please-but-never-mind-the-bread](https://realpython.com/primer-on-python-decorators/#both-please-but-never-mind-the-bread))

---
In reply to: [302770609](https://github.com/Azure/azure-iot-sdk-python-preview/pull/133#discussion_r302770609) [](ancestors = 302770609,302245692,301759558)